### ISSUE TITLE
`unique_together` constrain should let create a model if one of the envolved fields could be `blank`  

### ISSUE THREAD

**Issue by spac3-monkey** [tomchristie/django-rest-framework#2547]
When having a model with `unique_together` constrain involving two fields and one of the fields could be left `blank` it throws an error like "This field is required."

the code and test case here: http://pastebin.com/0HNqzJii

thanks in advance.

**Comment by tomchristie** [comment]
Assuming that this is a duplicate of #2452.

**Comment by tomchristie** [comment]
If you believe otherwise then file a more complete issue report and we can consider reopening.

**Comment by spac3-monkey** [comment]
The thing is that I'm not getting a uniqueness error. DRF ask for required field even when I specifically tells it to `blank=True` in model and `required=False` in serializer. 
Removing the constrain works perfectly.

Did you look at the code? I think it's not the same issue.

**Comment by tomchristie** [comment]
This is mentioned in [the documentation here](http://www.django-rest-framework.org/api-guide/validators/#uniquetogethervalidator)

> The UniqueTogetherValidation class always imposes an implicit constraint that all the fields it applies to are always treated as required. Fields with default values are an exception to this as they always supply a value even when omitted from user input.

The reason here is that we *need* to know the value that is being passed to the field at the point of the serializer in order to validate it. If you added the field explicitly on the serializer with `default=''` then you'll get the behavior you want.

Possible that we could improve the documentation around this, or look again at if there are ways we could improve the behavior eg automatically add `CreateOnlyDefault('')` on blank fields, so I'll leave this ticket open for a bit, as a reminder to take another look at this.

**Comment by pdr** [comment]
Just ran into the same issue. 

```
class LeadSerializer(ModelSerializer):
    slug = serializers.ReadOnlyField()
```

But in my model I have:
```
    class Meta:
        unique_together = ('slug', 'email')
```
Which made the serializer get the following validator:
```
   validators = [<UniqueTogetherValidator(queryset=Lead.objects.all(), fields=('slug', 'email'))>]
```

As a result it kept telling me 'slug': 'This field is required', but I had no idea the extra validator was getting added since it was auto generated from the model. It might be nice to raise a warning (i.e. you cannot use a readonly field with a UniqueTogetherValidator).

**Comment by MasterKale** [comment]
I experienced this exact same issue today. Here's the model I just added a `unique_together` constraint to:

    class Location(models.Model):
        company = models.ForeignKey(Company)
        title = models.CharField(max_length=100)

        class Meta:
            ordering = ['company']
            unique_together = ['company', 'title']

And here's its Serializer:

    class LocationSerializer(serializers.ModelSerializer):
        company = serializers.StringRelatedField()

        class Meta:
            model = Location

It seems that the Models' constraint was clashing with the Serializer's read-only `StringRelatedField`. As a result, whenever I try to create a Location using the Serializer:

    lo = LocationSerializer(data={'title': 'test', 'company': 2})

I receive back the following error:

    {'company': ['This field is required.']}

Prior to adding the `unique_together`, the serializer worked fine. In fact I hadn't touched this model or serializer in months and was really stumped as to why it just stopped working. 

The current error message is very misleading; it led me on a wild goose chase for an hour or so until I thought to check Github for similar issues. If this is Working As Intended<sup>tm</sup> then at the very least can we, like pdr suggested, get a different error message telling us not to use `unique_together` constraints with `read-only` fields?

**Comment by xordoquy** [comment]
@MasterKale The error message may look misleading though it's the most appropriate.
read-only fields are discarded for write operation meaning that you actually give {'title': 'test'} to the serializer. This leads to a validation error because there's no way you can create a model with only the title and the unique together constraint.
Am I missing a point here ?

**Comment by jamesdoherty** [comment]
@xordoquy I've run into a similar problem where one of the fields in the unique_together constraint is populated via a perform_create method. It doesn't appear possible to do that, so I've had to remove the constraint.

**Comment by xordoquy** [comment]
@jamesdoherty which confirms what I was saying.
Let me rephrase my previous comment.
You can't enforce `unique_together` constraint while one of the field is missing.
If your API makes one of the field optional, there will be no way to get a value for that field before validation is run.
Therefore there's no reason to have the `unique_together` in the serializer.

I think the developer should choose whether or not to remove the constraint.
I foresee to many use cases that make it hard to automate.
If the constraint is removed automatically by DRF because one of the field is optional, how will we cope with validation when that field is provided ?
What if one of the field can indeed have a None value ? It'll means we'll fail to catch the constraint during validation and end up with a 500 error.
And so on.

@tomchristie I'm willing to close this issue. My position would be that we could issue a warning at start time if we notice a `unique_together` constraint and a `read_only` or `required=False` constraint to help the developer notice there may be an issue.

**Comment by tomchristie** [comment]
> My position would be that we could issue a warning at start time if we notice a unique_together constraint and a read_only or required=False constraint to help the developer notice there may be an issue.

Gotcha - would be willing to consider that. Argument needs to be simple and coherent, and against a new issue *if* anyone wants to raise that.
